### PR TITLE
add an etcd client for naming resolution

### PR DESCRIPTION
**Pull request by yangzhouhan**

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -71,6 +71,13 @@ type dialOptions struct {
 // DialOption configures how we set up the connection.
 type DialOption func(*dialOptions)
 
+type NameResolver interface {
```
Comment (iamqizhao) [inline clientconn.go]: add comments for all exported type.

File: etcdnaming/etcdnaming.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	"sync"
+
+	etcdcl "github.com/coreos/etcd/client"
+)
+
+type namePair struct {
+	key, value string
+}
+
+// recvBuffer is an unbounded channel of item.
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: item -> *namePair

File: etcdnaming/etcdnaming.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		c: make(chan *namePair, 1),
+	}
+	return b
+}
+
+func (b *recvBuffer) put(r *namePair) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.backlog = append(b.backlog, r)
+	if !b.stopping {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: this check should be above the backlog op in line 33. Probably something like:

if b.stopping {
  return
}
b.backlog = append(...)
...

File: etcdnaming/etcdnaming.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+			b.backlog = b.backlog[1:]
+		default:
+		}
+	}
+}
+
+func (b *recvBuffer) load() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if len(b.backlog) > 0 && !b.stopping {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: if b.stopping || len(b.backlog) == 0 {
  return
}
...

File: etcdnaming/etcdnaming.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+func NewETCDNR(cfg etcdcl.Config) *etcdNR {
+	return &etcdNR{
+		cfg:  cfg,
+		recv: newRecvBuffer(),
+		ctx:  context.Background(),
+	}
+}
+
+func getNode(node *etcdcl.Node, res map[string]string) {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: add some comments.

File: etcdnaming/etcdnaming.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	recv   *recvBuffer
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewETCDNR(cfg etcdcl.Config) *etcdNR {
+	return &etcdNR{
+		cfg:  cfg,
+		recv: newRecvBuffer(),
+		ctx:  context.Background(),
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: you need to set cancel in this function too.

File: etcdnaming/etcdnaming.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	node := resp.Node
+	res := make(map[string]string)
+	getNode(node, res)
+	return res
+}
+
+func (nr *etcdNR) Watch(target string) {
+	cfg := nr.cfg
+	c, err := etcdcl.New(cfg)
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: can Watch and Get share the same "c"? If yes, we can use "c" to replace the "cfg" in etcdNR struct.

File: etcdnaming/etcdnaming.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+		fmt.Printf("non-nil error: %v", err)
+	}
+	node := resp.Node
+	res := make(map[string]string)
+	getNode(node, res)
+	return res
+}
+
+func (nr *etcdNR) Watch(target string) {
+	cfg := nr.cfg
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: not needed.

File: etcdnaming/etcdnaming.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		res[node.Key] = node.Value
+		return
+	}
+	for _, val := range node.Nodes {
+		getNode(val, res)
+	}
+}
+
+func (nr *etcdNR) Get(target string) map[string]string {
+	cfg := nr.cfg
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: not needed

File: etcdnaming/etcdnaming.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		if resp.Node.Dir {
+			continue
+		}
+		entry := &namePair{key: resp.Node.Key, value: resp.Node.Value}
+		nr.recv.put(entry)
+	}
+}
+
+func (nr *etcdNR) GetUpdate() (key string, val string) {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: comment the return value for ending.

Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: you do not need to have "key" and "val" in return.

**Comment by philips** [comment]
Is there a high-level description of how this would be used?

**Comment by iamqizhao** [comment]
Hi Philips,

A half-baked doc is at https://github.com/a11r/grpc/blob/doc2/doc/naming.md. And I will have a complementary doc for go specific.

File: clientconn.go [hunk 11]
```diff
@@ -71,6 +71,18 @@ type dialOptions struct {
 // DialOption configures how we set up the connection.
 type DialOption func(*dialOptions)
 
+// NameResolver defines the functions needed to resolve a customized names.
```
Comment (iamqizhao) [inline clientconn.go]: NameResolver does name resolution and watches for the resolution changes.

File: clientconn.go [hunk 12]
```diff
@@ -71,6 +71,18 @@ type dialOptions struct {
...
+type NameResolver interface {
+	// Get returns the map of customized names and the corresponding ip/DNS addresses.
```
Comment (iamqizhao) [inline clientconn.go]: Get gets a snapshot of the current name resolution results for target.

File: clientconn.go [hunk 13]
```diff
@@ -71,6 +71,18 @@ type dialOptions struct {
...
+	Get(target string) map[string]string
+	// Watch set a watcher on the target directory/key.
+	Watch(target string)
```
Comment (iamqizhao) [inline clientconn.go]: Watch watches for the name resolution changes on target. It blocks until Stop() is invoked. The watch results are obtained via GetUpdate().

File: clientconn.go [hunk 14]
```diff
@@ -71,6 +71,18 @@ type dialOptions struct {
...
+	// GetUpdate returns the key and value of the updated namePair of the target directory/key being watched.
+	GetUpdate() (string, string)
```
Comment (iamqizhao) [inline clientconn.go]: GetUpdate returns a name resolution change when watch is triggered. It blocks until it observes a change. The caller needs to call it again to get the next change.

File: clientconn.go [hunk 15]
```diff
@@ -71,6 +71,18 @@ type dialOptions struct {
...
+	// Stop shut down the watcher.
```
Comment (iamqizhao) [inline clientconn.go]: Stop shuts down the NameResolver.

File: etcdnaming/etcdnaming.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
+package etcdnaming
+
+import (
+	"fmt"
+	"golang.org/x/net/context"
+	"sync"
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: the order should be

fmt
sync

etcdcl ...
golang.org/x/net/contex

File: etcdnaming/etcdnaming.go [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+	etcdcl "github.com/coreos/etcd/client"
+)
+
+type namePair struct {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: namePair --> kv

**Comment by philips** [comment]
Thank you for the context, makes sense.

File: etcdnaming/etcdnaming.go [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+type etcdNR struct {
+	kAPI   etcdcl.KeysAPI
+	recv   *recvBuffer
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewETCDNR(cfg etcdcl.Config) *etcdNR {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: comment

File: etcdnaming/etcdnaming.go [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	return &etcdNR{
+		kAPI:   kAPI,
+		recv:   newRecvBuffer(),
+		ctx:    ctx,
+		cancel: cancel,
+	}
+}
+
+// getNode can be called recursively to build the result key-value map
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: getNode builds the resulting key-value map starting from node recursively.

File: etcdnaming/etcdnaming.go [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+		getNode(val, res)
+	}
+}
+
+func (nr *etcdNR) Get(target string) map[string]string {
+	resp, err := nr.kAPI.Get(nr.ctx, target, &etcdcl.GetOptions{Recursive: true, Sort: true})
+	if err != nil {
+		fmt.Printf("non-nil error: %v", err)
+	}
+	node := resp.Node
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: not needed

File: etcdnaming/etcdnaming.go [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+		}
+		entry := &kv{key: resp.Node.Key, value: resp.Node.Value}
+		nr.recv.put(entry)
+	}
+}
+
+func (nr *etcdNR) GetUpdate() (string, string) {
+	select {
+	case i := <-nr.recv.get():
+		nr.recv.load()string
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: remove string

File: etcdnaming/etcdnaming.go [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+type etcdNR struct {
+	kAPI   etcdcl.KeysAPI
+	recv   *recvBuffer
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewETCDNR creates an etcd NameResolver
+func NewETCDNR(cfg etcdcl.Config) *etcdNR {
```
Comment (iamqizhao) [inline etcdnaming/etcdnaming.go]: This needs to be restructured like:
naming/naming.go (type Resolver interface)
naming/etcd/etcd.go (etcd impl)

You can follow what credentials + oauth does as an example.

File: naming/etcd/etcd.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	res := make(map[string]string)
+	getNode(resp.Node, res)
+	return res
+}
+
+func (nr *etcdNR) Watch(target string) {
+	watcher := nr.kAPI.Watcher(target, &etcdcl.WatcherOptions{Recursive: true})
+	for {
+		ctx, cancel := context.WithCancel(nr.ctx)
```
Comment (iamqizhao) [inline naming/etcd/etcd.go]: nr.ctx and nr.cancel have been set in NewETCDNR already.

File: naming/etcd/etcd.go [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	for _, val := range node.Nodes {
+		getNode(val, res)
+	}
+}
+
+func (nr *etcdNR) Get(target string) map[string]string {
+	resp, err := nr.kAPI.Get(nr.ctx, target, &etcdcl.GetOptions{Recursive: true, Sort: true})
+	if err != nil {
+		fmt.Printf("non-nil error: %v", err)
```
Comment (iamqizhao) [inline naming/etcd/etcd.go]: log.Errorf("etcdNR.Get(_) failed %v", err)
return nil

File: naming/etcd/etcd.go [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+	etcdcl "github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/naming"
+)
+
+type kv struct {
+	key, value string
+}
+
+// recvBuffer is an unbounded channel of *kv.
```
Comment (iamqizhao) [inline naming/etcd/etcd.go]: // recvBuffer is an unbounded channel of *kv to record all the pending changes from etcd server.

File: naming/etcd/etcd.go [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+		b.backlog = b.backlog[1:]
+	default:
+	}
+}
+
+func (b *recvBuffer) get() <-chan *kv {
+	return b.c
+}
+
+func (b *recvBuffer) stop() {
```
Comment (iamqizhao) [inline naming/etcd/etcd.go]: // stop terminates the recvBuffer. After it is called, the recvBuffer is not usable any more.

File: naming/etcd/etcd.go [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+type etcdNR struct {
+	kAPI   etcdcl.KeysAPI
+	recv   *recvBuffer
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewETCDNR creates an etcd NameResolver
```
Comment (iamqizhao) [inline naming/etcd/etcd.go]: period at the end

File: naming/etcd/etcd.go [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+func NewETCDNR(cfg etcdcl.Config) naming.Resolver {
```
Comment (iamqizhao) [inline naming/etcd/etcd.go]: probably returning (name.Resolver, error) would be better.

**Pull request merged**